group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -0,0 +1,261 @@
+                    log.warn(memberId + " subscribed to topic " + topicId + " which doesn't exist in the topic metadata");
+                }
+            }
+        });
+
+        return membersPerTopic;
+    }
+
+    private Map<Uuid, List<Integer>> getUnassignedPartitionsPerTopic(final AssignmentSpec assignmentSpec, Map<Uuid, Set<Integer>> assignedStickyPartitionsPerTopic) {
+        Map<Uuid, List<Integer>> unassignedPartitionsPerTopic = new HashMap<>();
+        Map<Uuid, AssignmentTopicMetadata> topicsMetadata = assignmentSpec.topics();
+
+        topicsMetadata.forEach((topicId, assignmentTopicMetadata) -> {
+            ArrayList<Integer> unassignedPartitionsForTopic = new ArrayList<>();
+            int numPartitions = assignmentTopicMetadata.numPartitions();
+            // List of unassigned partitions per topic contains the partitions in ascending order.
+            Set<Integer> assignedStickyPartitionsForTopic = assignedStickyPartitionsPerTopic.getOrDefault(topicId, new HashSet<>());
+            for (int i = 0; i < numPartitions; i++) {
+                if (!assignedStickyPartitionsForTopic.contains(i)) {
+                    unassignedPartitionsForTopic.add(i);
+                }
+            }
+            unassignedPartitionsPerTopic.put(topicId, unassignedPartitionsForTopic);
+        });
+
+        return unassignedPartitionsPerTopic;
+    }
+
+/**
+ * <p> The algorithm includes the following steps:
+ * <ol>
+ * <li> Generate a map of <code>membersPerTopic</code> using the given member subscriptions.</li>
+ * <li> Generate a list of members (<code>potentiallyUnfilledMembers</code>) that have not met the minimum required quota for assignment AND
+ * get a list (<code>assignedStickyPartitionsPerTopic</code>) of partitions that we want to retain in the new assignment.</li>
+ * <li> Add members from the <code>potentiallyUnfilledMembers</code> list to the <code>unfilledMembers</code> list if they haven't met the total required quota i.e. minimum number of partitions per member + 1 (if member is designated to receive one of the excess partitions). </li>
+ * <li> Generate a list of unassigned partitions by calculating the difference between total partitions and already assigned (sticky) partitions. </li>
+ * <li> Iterate through the unfilled members and assign partitions from the unassigned partitions. </li>
+ * </ol>
+ * </p>
+ */
+    @Override
+    public GroupAssignment assign(final AssignmentSpec assignmentSpec) throws PartitionAssignorException {
+        Map<String, Map<Uuid, Set<Integer>>> membersWithNewAssignmentPerTopic = new HashMap<>();
+        // Step 1
+        Map<Uuid, List<String>> membersPerTopic = membersPerTopic(assignmentSpec);
+        // Step 2
+        Map<Uuid, List<RemainingAssignmentsForMember<String, Integer>>> unfilledMembersPerTopic = new HashMap<>();
+        Map<Uuid, Set<Integer>> assignedStickyPartitionsPerTopic = new HashMap<>();
+
+        membersPerTopic.forEach((topicId, membersForTopic) -> {
```

Comment:
I still wonder if we could combine steps 4 and 5 in this loop. For instance, could we do something like this?

* We start by creating a sorted set with all the partitions of the topic.
* Then for each member, we do what is already done but instead of populating `assignedStickyPartitionsPerTopic`, we remove assigned partitions from the sorted set.
* Then we go through the unfilled members and allocated the remaining partitions in the sorted set.

This could potentially reduce the number of data structures.

Reply:
Steps 3 & 4?

Reply:
We would need to have a Map<Uuid, SortedSet> called partitions and to remove every assigned partition each removal would cost O(logn) so for n removals worst case O(nlogn). The time complexity of just calculating at the end is O(n). So its really just time vs space. we save space of one map O(n+m) in this method but we spend more time for removal.

Reply:
2maps and O(n) time VS 1map (potentially takes up more space if its a tree set for sorted order) and O(nlogn) time

Reply:
I remember now actually, initially I did add it inside the same loop but I got comments to make the code more modular so I made a new function to get unassigned partitions :( I changed it now but yeah that's what happened

Reply:
I think you mean step 4? step 3 is filling in the potentially unfilled members map and that i can't elminate.